rfcs/0097-unify-packed-and-object.md

```diff
@@ -0,0 +1,677 @@
+
+The unified List can be used as underlying container for specialized data structures (e.g. List<T>):
+
+- UnifyItem to T:
+    - Depends on its type, either use Object to T, or AnyValue to T.
+- set T: turn T into UnifyItem:
+    - If T is object
+        - turn into Object
+        - If T is boxed object, turn into Pod
+    - If T is pod, turn into Pod
+
+**D3c-variant, further reducing cost of small int and char**
+
+We can further reduce the cost of memory cost in D3c by having a static AnyValue table for common small integers and frequently appearing characters. Specifically, we can allocate a static part of AnyValue pool for integer in a small range and not allocate from local AnyValue pool. 
+
+**Summary of tradeoffs:**
+
+Memory cost:
+
+- D3a: 24 bytes per Object, 24 bytes per POD(small-str)
+- D3b: 8 byte per Object, 8 bytes ptr + 24 bytes(boxed object) if original value is not boxed.
+- D3c: 8 bytes per Object, 32 bytes per POD(small-str)
+- D3c-variant: 8 bytes per Object, 8 bytes per small integer and char that have static pool.
+
+Accessing efficiency:
+
+D3b may have overhead for small-str and POD. D3a and D3c are small-str friendly.
+
+### D4: PackedFunc Convention
+
+In this section, we revisit the PackedFunc convention under the new context. The update to the C++ PackedFunc API will run as follows:
+
+- TVMArgs now stores ptr of AnyView, TVMArgs[i] will now return AnyView
+- Alias: TVMRetValue = AnyValue, TVMArgValue = AnyView
+
+These changes are invisible to the users as long as they use the same source library. We will immiedately gain the ability to do universal switching when defining PackedFunc.
+
+- A developer can choose to develop solely on ObjectRef, in this case automatic conversion happens when we turn AnyView and request Object. We expect most of the compiler development to be in this mode.
+- A developer can choose to develop runtime functions that contains AnyValue and AnyView to take benefit of stack values. This will have benefit that intermediate values can store small-str. We anticipate some of the builtin runtime to operate on this mode.
+- String will be backed by both small-str and object for efficiency, which we expect to help compiler as well(as there are a lot of small names).
+- Both runtime and compiler will be backed by universal container object, which allows us to simplify the automatic conversion in FFI (you take a python tuple and would expect an Array).
+
+There are several design choices in terms of C API convention in light of this new proposal. They will affect the internal data layout of TVMArgs.
+
+- **D4a:** Current C API:
+    
+    ```cpp
+    int TVMCPackedFunc(PackedFuncHandle handle,
+                       int num_args, int* type_codes, TVMValue* values,
+                       int* out_tcode, TVMValue* out_value);
+    ```
+    
+    - This would require packing unpacking the typecode. Note that small string passing wont work because of the lack of the seq_len padding field.
```

Comment:
```suggestion
    - This would require unpacking the typecode. Note that small string passing wont work because of the lack of the seq_len padding field.
```